pg_shard.c

```diff
@@ -1004,6 +1407,215 @@ TargetEntryList(List *expressionList)
 
 
 /*
+ * CreateTemporaryTableLikeStmt returns a CreateStmt node depending on whether the
+ * aggregates in the localQuery can be pushed down to the workers or not.
+ */
+static CreateStmt *
+CreateTemporaryTableStmt(Query *distributedQuery, bool pushDownAggregates)
+{
+	CreateStmt *createStmt = NULL;
+
+	if (pushDownAggregates)
+	{
+		List *workerTargetList = distributedQuery->targetList;
+		List *rangeTableList = distributedQuery->rtable;
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
+		List *columnNameValueList = rangeTableEntry->eref->colnames;
+		List *columnNameList = ValueToStringList(columnNameValueList);
+		List *targetList = MasterTargetList(workerTargetList);
+		StringInfo tempTableName = CreateUniqueTableName();
+
+		createStmt = BuildCreateStatement(tempTableName, targetList, columnNameList);
+	}
+	else
+	{
+		Oid distributedTableId = ExtractFirstDistributedTableId(distributedQuery);
+
+		createStmt = CreateTemporaryTableLikeStmt(distributedTableId);
+	}
+
+	return createStmt;
+}
+
+
+/*
+ * ValueToStringList walks over the given list of string value types, converts
+ * value types to cstrings, and adds these cstrings into a new list.
+ */
+static List *
+ValueToStringList(List *valueList)
```

Comment:
I don't like the name of this method (or its comment), but seeing as how it comes from within CitusDB I think we must leave it.